### PR TITLE
Swap promises for await

### PR DESCRIPTION
**Pull request by michaelprichardson**
Replaced the promise arrays with await/async. This helped to remove the `sleep()` in the tests

### CODE REVIEW

File: src/rules/deleteReferences.ts [hunk 1]
```diff
@@ -107,19 +114,15 @@ export function integrifyDeleteReferences(
...
-                `integrify: Deleting [${target.collection}]${
-                  target.isCollectionGroup ? ' (group)' : ''
-                }, id [${doc.id}]`
-              );
-              promises.push(doc.ref.delete());
-            });
-          })
-        );
-      });
-      return Promise.all(promises);
```
Comment (nbransby) [inline src/rules/deleteReferences.ts:123]: Might still be worth using Promise.all to get these queries to run in parallel as now they will be sequential.

File: src/rules/deleteReferences.ts [hunk 2]
```diff
@@ -107,19 +114,15 @@ export function integrifyDeleteReferences(
...
+        const querySnap = await whereable.get();
+        for (const doc of querySnap.docs) {
+          console.log(
+            `integrify: Deleting [${target.collection}]${
+              target.isCollectionGroup ? ' (group)' : ''
+            }, id [${doc.id}]`
+          );
+          await doc.ref.delete();
```
Comment (nbransby) [inline src/rules/deleteReferences.ts]: it deletes each record one by one? Should really use a WriteBatch here

File: src/rules/replicateAttributes.ts [hunk 3]
```diff
@@ -121,26 +121,19 @@ export function integrifyReplicateAttributes(
...
+        for (const doc of detailDocs.docs) {
+          console.log(
+            `integrify: On collection ${
+              target.isCollectionGroup ? 'group ' : ''
+            }[${target.collection}], id [${doc.id}], applying update:`,
+            update
+          );
+          await doc.ref.update(update);
+        }
+      }
```
Comment (nbransby) [inline src/rules/replicateAttributes.ts]: same points apply here

File: src/common.ts [hunk 4]
```diff
@@ -10,6 +17,11 @@ export interface Config {
   };
 }
 
+export type PreHookFunction = (
+  change: DocumentSnapshot | Change<QueryDocumentSnapshot>,
+  context: functions.EventContext
+) => Promise<void>;
```
Comment (nbransby) [inline src/common.ts]: Consider making the return type ` Promise<void> | void` that way clients dont have to use the async keyword if they usign await

**Review by nbransby — COMMENTED** [review]
looks good. you might want to copy our WriteBatch class in and use that instead as this will blow up if there is more than 500 records in the collection

**Comment by michaelprichardson** [comment]
Added the WriteBatch class and replaced the standard batch writes in delete and replicate. There was an issue with the import of `object-sizeof`, had to replace `sizeof` with `require('object-sizeof')`

File: test/unit.test.js [hunk 5]
```diff
@@ -409,6 +409,7 @@ async function testDeleteReferences(sut, t, name) {
     t.is(state.context.params.masterId, masterId);
   }
 
+  console.log('Here 1');
```
Comment (nbransby) [inline test/unit.test.js]: guessing this can be removed now

File: test/functions/index.js [hunk 6]
```diff
@@ -33,7 +33,7 @@ module.exports.replicateMasterToDetail = integrify({
     },
   ],
   hooks: {
-    pre: (change, context) => {
+    pre: async (change, context) => {
```
Comment (nbransby) [inline test/functions/index.js]: can async be removed from here now or is setState async?

**Review by nbransby — APPROVED** [review]

**Pull request merged**